File: backend/app/api/projects.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import require_permission
from app.core.database import get_db
from app.models.project import Project
from app.models.user import User
from app.schemas.project import ProjectCreate, ProjectOut, ProjectUpdate

router = APIRouter(prefix="/projects", tags=["projects"])


@router.get("", response_model=list[ProjectOut])
async def list_projects(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("projects", "view")),
):
    result = await db.execute(select(Project).order_by(Project.id.desc()))
    return result.scalars().all()


@router.post("", response_model=ProjectOut, status_code=status.HTTP_201_CREATED)
async def create_project(
    payload: ProjectCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("projects", "create")),
):
    project = Project(**payload.model_dump())
    db.add(project)
    await db.commit()
    await db.refresh(project)
    return project


@router.patch("/{project_id}", response_model=ProjectOut)
async def update_project(
    project_id: int,
    payload: ProjectUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("projects", "edit")),
):
    project = await db.get(Project, project_id)
    if project is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Loyiha topilmadi")

    for field, value in payload.model_dump(exclude_unset=True).items():
        setattr(project, field, value)

    await db.commit()
    await db.refresh(project)
    return project


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_permission("projects", "delete")),
):
    project = await db.get(Project, project_id)
    if project is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Loyiha topilmadi")
    await db.delete(project)
    await db.commit()

File: admin-panel/src/api/projects.ts
import { apiClient } from "./client";
import type { ProjectOut } from "../types";

export interface ProjectPayload {
  name: string;
  description?: string | null;
  status?: string;
  responsible_person?: string | null;
}

export async function listProjects(): Promise<ProjectOut[]> {
  const { data } = await apiClient.get<ProjectOut[]>("/projects");
  return data;
}

export async function createProject(payload: ProjectPayload): Promise<ProjectOut> {
  const { data } = await apiClient.post<ProjectOut>("/projects", payload);
  return data;
}

export async function updateProject(id: number, payload: Partial<ProjectPayload>): Promise<ProjectOut> {
  const { data } = await apiClient.patch<ProjectOut>(`/projects/${id}`, payload);
  return data;
}

export async function deleteProject(id: number): Promise<void> {
  await apiClient.delete(`/projects/${id}`);
}
